### PR TITLE
Fix: code blocks not keyboard focusable

### PR DESCRIPTION
**Pull request by jcarstairs-scottlogic**
TODO:

* ensure all comments are sufficiently addressed
* put up a preview deployment

Before merging, carry out these [important steps](#important-steps)!

This PR makes code blocks usable to people who use a keyboard as their primary input modality. As a result of the technical requirements for this, it also replaces our current, plug-and-play build-and-deploy process with a custom GitHub Actions workflow.

## The problem: code blocks not keyboard-focusable

Many blog posts contain code blocks. Some of these are wide enough that you need to scroll horizontally to read all the content. (If a user happens to increase font size, for example, because they have limited vision, this will be true of more code blocks for them.)

In order for users who navigate primarily with the keyboard to be able to scroll these code blocks horizontally, they need to be present in the tab order. However, on Chrome and Edge (not sure about Safari) code blocks are not present in the tab order by default. To fix this, you have to attach the `tabindex` attribute yourself with a value greater than or equal to `0`.

## The solution: a custom Jekyll plugin

This PR implements a custom Jekyll plugin, [make_code_blocks_keyboard_focusable.rb][make-code-blocks-keyboard-focusable]. It takes in every post after it has been converted to HTML, picks out every `<pre>` element (which always wraps `<code>` blocks in order to preserve whitespace), and, if the `tabindex` attribute isn't already present, sets it to `0`. This means you can focus code blocks with the keyboard and scroll them horizontally with the arrow keys, no matter which browser you're using.

### Are there better solutions?

- If someone had already written this plugin, that would have saved us some effort. But I can't find any such plugin. (In any case, writing the plugin is not the burden of the effort for this change, and we've already done it now.)
- You could add the `tabindex` manually. But this has several drawbacks:
  - Most authors don't write code themselves, they use SiteLeaf to generate Markdown. We can't control how SiteLeaf generates Markdown.
  - Realistically, even when authors write their own code, authors are simply unlikely to add the `tabindex`. It's not obvious, so they'd need to read some instructions to tell them to do it. But nobody reads instructions, ever. Even if they did, they'd need to remember to do it when they actually got round to writing the code. But they would often forget.
  - Even when authors write their own code, and they read the instructions, and they remember, if they're writing Markdown, it is technically impossible. They can add a `tabindex`, but it'll go on a `div` wrapper around the code block, and this won't solve the problem of being able to scroll the code block (@ithake has tested this).
- You could write a Markdown preprocessor which adds `tabindex` to code blocks. But:
  - This has the same problem mentioned before, that the `tabindex` goes on a `div` wrapper, which means the problem isn't solved.
  - It wouldn't fix HTML blog posts anyway.

## Why this affects our build-and-deploy process

With GitHub Pages, there are two kinds of "publishing source": you can "publish from a branch" or write a custom GitHub Actions workflow file. (See [Pages docs on publishing sources][pages-docs-publishing-sources].) Right now, we "publish from a branch", but to implement our custom plugin, we need to change to the custom workflow file approach.

### Costs vs benefits of changing our build process

Costs:

- Theoretically implies a maintenance cost for the workflow file. (In practice, we have few dependencies and those we do have have stable APIs, so I don't think there's likely to be a significant maintenance cost)

Benefits:

- Can get these code blocks to be accessible to keyboard users!
- From now on, can install whatever plugins we like
- From now on, we're in control of our build process, so if we want to change it in any way in future, we can

_NB. This should have no impact on authors. I've already tested this out with manual authoring anyway, but there's no reason to expect that SiteLeaf authors will be affected, either. What the pipeline does after they've committed their changes is not the author's problem._

### Publish from a branch, and why this isn't compatible with a custom plugin

The simplest publishing source is to "publish from a branch", as we do currently. This means you just push a Jekyll project to the `gh-pages` branch and some GitHub magic builds and deploys the website for you.

Under the hood, the build step for this is carried out by the [jekyll-build-pages][jekyll-build-pages] Action, which depends on the `github-pages` gem.

Meanwhile, the fix in this PR requires a custom plugin. In order to use custom plugins, you cannot run Jekyll in "safe mode". (For all that that makes the alternative sound "unsafe", it doesn't obviously have any dangerous consequences -- see [Jekyll configuration options docs][jekyll-docs-options] to get the full details on what this entails.)

But the `github-pages` gem is not compatible with unsafe mode. So when the jekyll-build-pages Action runs, since it depends on `github-pages`, it forces Jekyll to run in safe mode, which means it doesn't use our custom plugin.

Instead, we have to use the second "publishing source". That is, we have to write our own GitHub Action workflow file to write our own custom build-and-deploy workflow. That's what I've done in this PR.

### The custom build and deploy workflow

This does exactly what you would expect: it builds the project with `bundle exec jekyll build`, uploads the result and deploys it to GitHub Pages.

You can see it in action in [my fork][demo-build-and-deploy-workflow]! It deploys to [my personal GitHub Pages domain][demo-blog-deployment]. The workflow file is `build-and-deploy.yaml`.

<h2 id="important-steps">Important: steps to merge this PR if it's been approved</h2>

1. Go to repository Settings > Pages > Build and deployment.
2. Change the deployment "Source" from "Deploy from a branch" to "GitHub Actions". *
3. Set the "Custom domain" to "blog.scottlogic.com". **
4. Merge this PR.

\* Setting the publishing source will prevent GitHub from trying to deploy the blog from its automatic workflow as well as our custom one.
** If you don't explicitly set the custom domain here, it'll deploy to something like `scottlogic.github.io/blog`.

[demo-blog-deployment]: https://jcarstairs-scottlogic.github.io/blog
[demo-build-and-deploy-workflow]: https://github.com/jcarstairs-scottlogic/blog/deployments
[jekyll-build-pages]: https://github.com/actions/jekyll-build-pages
[jekyll-docs-options]: http://jekyllrb.com/docs/configuration/options/
[make-code-blocks-keyboard-focusable]: https://github.com/ScottLogic/blog/compare/gh-pages...ithake:blog-slogic:fix/code-blocks-not-keyboard-focusable?expand=1#diff-6095e9dfff7671b43549e5c87c20f1151ee969b8fd90989c1466d8588053f3e9
[pages-docs-publishing-sources]: https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#publishing-sources-for-github-pages-sites

### CODE REVIEW

**Comment by ithake** [comment]
Lot of content here - given what we would like Colin to consider, it might be better as an email expressing the problem, with a link to this PR as an example proof of concept (and link to a document repeating the details of the problem!) 

TLDR;
Applies to: Posts that use markdown to define code snippets.

An important part of accessibility is letting keyboard-only users view the contents of the site.
By design, example code snippets should maintain their line breaks so that a very long line in the code is rendered as a very long line within the browser, that the user will use a horizontal scroll bar to fully view.
Unfortunately an ongoing browser feature is that by default `<pre>` blocks fall outside the tab index order, so that a keyboard user cannot tab to move their focus to the code block.  Without that focus the user cannot access the scroll bar.

Due to the limitations of the available markdown processors (see details below), any tabindex attributes will need to be generated as HTML either pre- or post- the existing markdown processing.

This PR is an example of post-markdown processing.  Because it requires a plugin inconsistent with gh-pages, it will require a change to our existing "publish from a branch" process for authors.

**Comment by ithake** [comment]
### Limitations of Markdown processors ###

#### Target HTML
```
<pre tabindex="0">
  <code>
   Lorem ipsum dolor sit amet, ...
  </code>
</pre>
```

In effect we want to apply the `tabindex` attribute to the `pre` element.

#### Github Flavoured Markdown
Built on CommonMark. No mechanism for passing attributes from within markdown.

#### Kramdown
Permits [assigning attributes to blocks](https://kramdown.gettalong.org/quickref.html#block-attributes).
Accordingly `{:tabindex="0"}` is legal in kramdown.

Unfortunately code blocks behave slightly differently.

Given:
``````
{:tabindex="0"}
~~~html 
Lorem ipsum dolor sit amet, ...
~~~
``````
The generated HTML contains additional `div` elements:
```
<div tabindex="0" class="language-html highlighter-rouge">
  <div class="highlight">
    <pre class="highlight">
      <code>
Lorem ipsum dolor sit amet, ...
      </code>
    </pre>
  </div>
</div>
```
Here, although an ancestor is assigned the provided attribute, this is still separate from (the width of?) the `pre` element scroll bar and there is no way to tab between the two.

(Also tested for different position of the tabindex definition, before the block, after the block, on the same line as the block).

**Comment by ithake** [comment]
## Pre-processor options... ##
Options/reasons for rejection be documented separately?
- presumably would also require gh-pages/incompatible plugin, or could you sneak in an earlier filter written in node to preprocess the markdown code block elements only?
- I'm aware that's getting near the point where you might start just switching to a different static site generation engine altogether!

**Comment by ithake** [comment]
1. FYI - general advice is to consistently set the tabindex at 0 so that the tab order matches the order within the source code.
1. FYI "-1" is also syntactically valid, but is used to actively exclude the element from the default tab index.
1. FYI - in 2015, in response to a different [tabindex issue](https://github.com/gettalong/kramdown/issues/215#issuecomment-73093711), the originator of kramdown  emphasised post-processing as the best way forward, as per the approach here.

File: _plugins/make_code_blocks_keyboard_focusable.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+require 'nokogiri'
+
+Jekyll::Hooks.register :documents, :post_render do |document|
+  Jekyll.logger.debug "Making code blocks keyboard focusable in #{document.relative_path}..."
+  count_code_blocks = 0
+
+  html = Nokogiri.HTML5(document.output)
+  html.css('pre').each do |pre|
+    pre['tabindex'] = '0' unless pre.has_attribute?('tabindex')
```
Comment (ithake) [inline _plugins/make_code_blocks_keyboard_focusable.rb:9]: It's hard for me to think of a reason why an author would be explicitly adding a tabindex in their content, so this is probably good enough, but as this behaviour could surprise them, we might want to add something in the documentation for them.

**Comment by ithake** [comment]
> Instead, we have to use the second "publishing source". That is, we have to write our own GitHub Action workflow file to write our own custom build-and-deploy workflow. That's what I've done in this PR.

What does that mean for an author/site-leaf user?  Is there any change to what they do?

**Comment by jcarstairs-scottlogic** [comment]
> > Instead, we have to use the second "publishing source". That is, we have to write our own GitHub Action workflow file to write our own custom build-and-deploy workflow. That's what I've done in this PR.
> 
> What does that mean for an author/site-leaf user? Is there any change to what they do?

Good question.

* There's no reason why the flow should change for users, whether manual or SiteLeaf. Authors make changes and commit them. Pipelines take committed changes and do whatever they please with them (in this case build & deploy).
* That's the theory, I'm pretty confident of it.
* It's also the practice for manual changes, I've tested it.
* As for SiteLeaf, I don't think it is testable, but I don't think that should worry us. I guess SiteLeaf is hooked up specifically to the `gh-pages` branch, so there probably isn't a way to check if it still works before merging. But -- it's unlikely to break, and even if it does, the harm is minimal: someone can't commit a new blog post for a day or two until we fix it.

File: _plugins/make_code_blocks_keyboard_focusable.rb (hunk 1 above)
Comment (jcarstairs-scottlogic) [reply _plugins/make_code_blocks_keyboard_focusable.rb:9]: I think this case is improbable enough that we don't need to worry about it
